proto/config_test.go

```diff
@@ -151,3 +152,83 @@ func TestPermConfig(t *testing.T) {
 		t.Errorf("unexpected read access for user \"bar\"")
 	}
 }
+
+func verifyOrdering(attrs []string, rs ReplicaSlice, prefixLen int) bool {
+	prevMatchIndex := len(attrs)
+	for i := range rs {
+		matchIndex := -1
+		for j := range attrs {
+			if j >= len(rs[i].Attrs.Attrs) || rs[i].Attrs.Attrs[j] != attrs[j] {
+				break
+			}
+			matchIndex = j
+		}
+		if matchIndex != -1 && matchIndex > prevMatchIndex {
+			return false
+		}
+		if i == 0 && matchIndex+1 != prefixLen {
+			return false
+		}
+		prevMatchIndex = matchIndex
+	}
+	return true
+}
+
+func TestReplicaSetSortByCommonAttributePrefix(t *testing.T) {
+	replicaAttrs := [][]string{
+		[]string{"us-west-1a", "gpu"},
```

Comment:
Don't change anything now, but we should think about having several classifications of attributes. We don't always want to choose the nodes that have the most in common with the gateway node.
* Location: prefer the nodes that are most similar to the gateway node
* Storage type: e.g. prefer ssd over hdd no matter what the gateway node is using (perhaps with an option for low-priority mapreduces to prefer hdd to reduce load on ssds)
* Correlated failures (e.g. pdu): irrelevant to DistSender, but used in replica assignment
* Special attributes (e.g. gpu): use only when the query can take advantage of it